README.md
See https://github.com/linuxmuster/linuxmuster-opsi/wiki.


debian/changelog
linuxmuster-opsi (1.5.0-0ubuntu0) opsi; urgency=low

  * added support for dynamic ip clients.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Sat, 18 Apr 2020 22:04:24 +0200

linuxmuster-opsi (1.4.0-0ubuntu0) opsi; urgency=low

  * added helper script opsiSetup.py to restore opsi product client state and force
    certain products to setup if requested after linbo sync.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Thu, 14 Mar 2019 09:51:05 +0100

linuxmuster-opsi (1.3.3-0ubuntu0) opsi; urgency=low

  * added upgrade script.
  * removed lmn62 code.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Mon, 11 Mar 2019 22:07:18 +0100

linuxmuster-opsi (1.3.2-0ubuntu0) opsi; urgency=low

  * prepare: sets autoInstall = true in uib-windows.repo.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Mon, 11 Mar 2019 15:12:05 +0100

linuxmuster-opsi (1.3.1-0ubuntu0) opsi; urgency=low

  * more modifications due to opsi version 4.1 and ubuntu version 18.04.
  * removed last remnants of network setup.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Fri, 08 Mar 2019 21:38:12 +0100

linuxmuster-opsi (1.3.0-0ubuntu0) opsi; urgency=low

  * switched to opsi version 4.1 and ubuntu version 18.04.
  * removed network and hostname setup (now done by linuxmuster-prepare).

 -- Thomas Schmitt <thomas@linuxmuster.net>  Tue, 22 Jan 2019 15:05:23 +0100

linuxmuster-opsi (1.2.1-0ubuntu0) opsi; urgency=low

  * updated uib repository paths.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Mon, 07 May 2018 10:11:42 +0200

linuxmuster-opsi (1.2.0-0ubuntu0) opsi; urgency=low

  * environment, set_opsipassword(): added option to set password per cli parameter.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Tue, 01 May 2018 10:37:56 +0200

linuxmuster-opsi (1.1.1-0ubuntu0) opsi; urgency=low

  * deactivated email notification in opsi-product-updater.conf.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Mon, 26 Feb 2018 22:28:25 +0100

linuxmuster-opsi (1.1.0-0ubuntu0) opsi; urgency=low

  * Added support for lmn7.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Fri, 16 Feb 2018 13:12:31 +0100

linuxmuster-opsi (1.0.5-0ubuntu0) opsi; urgency=low

  * Added support for Ubuntu 16.04, removed support for 12.04
    (https://github.com/linuxmuster/linuxmuster-opsi/issues/1)
  * Added cronjob for opsi-product-updater
    (https://github.com/linuxmuster/linuxmuster-opsi/issues/2)
  * Added network configuration option to linuxmuster-opsi.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Tue, 18 Jul 2017 10:45:55 +0200

linuxmuster-opsi (1.0.4-0ubuntu1) opsi; urgency=low

  * Rebuild.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Thu, 26 Feb 2015 12:30:13 +0100

linuxmuster-opsi (1.0.4-0ubuntu0) opsi; urgency=low

  * fixed missing ip update for config.ini in setup.sh.
  * added ubuntu 14.04 support.
  * opsi-product-updater.conf template: activated wol feature.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Thu, 26 Feb 2015 12:26:58 +0100

linuxmuster-opsi (1.0.3-0ubuntu0) opsi; urgency=low

  * linuxmuster-opsi --prepare: improved dist-upgrade.
  * opsi-product-updater.conf: excluded package uedit32 in template.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Tue, 18 Mar 2014 17:49:16 +0100

linuxmuster-opsi (1.0.2-1ubuntu0) opsi; urgency=low

  * linuxmuster-opsi --prepare: removed update of opsi-product-updater.conf again.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Mon, 17 Mar 2014 19:57:47 +0100

linuxmuster-opsi (1.0.2-0ubuntu0) opsi; urgency=low

  * linuxmuster-opsi --prepare: fixed missing update of opsi-product-updater.conf.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Mon, 17 Mar 2014 17:49:24 +0100

linuxmuster-opsi (1.0.1-0ubuntu0) opsi; urgency=low

  * added dependecy to openbsd-inetd.
  * added setup for atftpd inetd start.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Mon, 17 Mar 2014 09:04:54 +0100

linuxmuster-opsi (1.0.0-0ubuntu0) opsi; urgency=low

  * initial release.

 -- Thomas Schmitt <thomas@linuxmuster.net>  Wed, 05 Mar 2014 16:34:17 +0100


share/setup.sh
#
# linuxmuster-opsi-setup
#
# thomas@linuxmuster.net
# 20190308
#

# read linuxmuster.net environment
. "$SETTINGS" || exit 1
# mailip is serverip on lmn62
grep -q ^mailip "$SETTINGS" || mailip="$serverip"

# test if necessary pkgs are installed
pkgs_installed &> /dev/null || bailout "The neccessary opsi packages are not installed!"

# get network stuff
for i in $OPSICONFIGDIR/depots/*.ini; do
 MYFQDN="$(basename "$i" | sed -e 's|.ini||')"
 [ -n "$MYFQDN" ] && break
 MYFQDN=""
done
[ -z "$MYFQDN" ] && MYFQDN="$(grep ^127 /etc/hosts | grep -v localhost | awk '{ print $2 }' | head -1)"
[ -z "$MYFQDN" ] && MYFQDN="$(hostname -f 2> /dev/null)"
[ -z "$MYFQDN" ] && bailout "Cannot get hostname!"
MYHOSTNAME="${MYFQDN%%.*}"
MYDOMAIN="${MYFQDN/$MYHOSTNAME./}"
MYIP="$(ip route get 8.8.8.8 | head -1 | awk '{ print $7 }')"
[ -z "$MYIP" ] && bailout "Cannot get ip address!"
[ "$MYIP" = "$opsiip" ] || bailout "Ip address is not $opsiip!"
MYNETMASK="$(ifconfig | grep -w inet | grep "$MYIP" | awk '{print $4}' | cut -d ":" -f 2)"
MYNETWORK="$(ipcalc $MYIP/$MYNETMASK | grep ^Network: | awk '{ print $2 }' | awk -F\/ '{ print $1 }')"

# control variables
cert_ok=yes
rand_conf="/var/tmp/opsirand.cnf.$$"
openssl_conf="/var/tmp/opsissl.cnf.$$"
[ -n "$first" ] && cert_ok=no


# functions begin

set_fqdn(){
 local RC="0"
 if [ -e "$PCKEYS" ]; then
  sed -i "$PCKEYS" -e "s|$MYFQDN|opsi.$domainname|g" -e "s|\..*\:|\.$domainname\:|g" || RC="1"
 else
  RC="1"
 fi
 # config.ini
 if [ -e "$OPSICONFIG" ]; then
  sed -e "s|$MYFQDN|opsi.$domainname|g" -i "$OPSICONFIG" || RC="1"
 else
  RC="1"
 fi
 # depot_ini
 local depot_ini="$OPSICONFIGDIR/depots/$MYFQDN.ini"
 if [ -e "$depot_ini" ]; then
  sed -e "s|$MYFQDN|opsi.$domainname|g
          s|^remoteurl .*|remoteurl = smb:\/\/opsi\/opsi_depot|g
          s|^ipaddress .*|ipaddress = $opsiip|g
          s|^network .*|network = $MYNETWORK/$MYNETMASK|g" -i "$depot_ini" || RC="1"
  mv "$depot_ini" "$(dirname $depot_ini)/opsi.$domainname.ini"
 else
  RC="1"
 fi
 # clients
 local i
 local newname
 if ls "$OPSICLIENTSDIR"/*.ini &> /dev/null; then
  for i in "$OPSICLIENTSDIR"/*.ini; do
   new_ini="$(basename "$i" | awk -F\. '{ print $1 }').$domainname.ini"
   mv "$i" "$(dirname "$i")/$new_ini" || RC="1"
  done
 fi
 return "$RC"
}

set_ip(){
 # write ip to config.ini
 local RC=0
 if [ -e "$OPSICONFIG" ]; then
  sed -e "s|https://.*|https://${opsiip}:4447/rpc\"\]|g" -i "$OPSICONFIG" || RC="1"
 else
  RC="1"
 fi
 return "$RC"
}

write_cert_data_to_debconf(){
 local RC="0"
 echo "set opsiconfd/cert_country $country" | debconf-communicate -f noninteractive || RC="1"
 echo "set opsiconfd/cert_state $state" | debconf-communicate -f noninteractive || RC="1"
 echo "set opsiconfd/cert_locality $location" | debconf-communicate -f noninteractive || RC="1"
 echo "set opsiconfd/cert_organization $schoolname" | debconf-communicate -f noninteractive || RC="1"
 echo "set opsiconfd/cert_unit linuxmuster.net" | debconf-communicate -f noninteractive || RC="1"
 echo "set opsiconfd/cert_commonname opsi.${domainname}" | debconf-communicate -f noninteractive || RC="1"
 echo "set opsiconfd/cert_email ${admin}@${domainname}" | debconf-communicate -f noninteractive || RC="1"
 return "$RC"
}

write_cert_tmp_config(){
 local RC
 dd if=/dev/urandom of="$rand_conf" count=1 2>/dev/null
 cat << EOF > "$openssl_conf"
RANDFILE = $rand_conf

[ req ]
default_bits = 1024
encrypt_key = yes
distinguished_name = req_dn
x509_extensions = cert_type
prompt = no

[ req_dn ]
C=$country
ST=$state
L=$location
O=$schoolname
OU=linuxmuster.net
CN=opsi.${domainname}
emailAddress=${admin}@${domainname}

[ cert_type ]
nsCertType = server
EOF
 RC="$?"
 return "$RC"
}

create_opsi_cert(){
 local RC="0"
 write_cert_data_to_debconf &> /dev/null || RC="1"
 write_cert_tmp_config || RC="1"
 openssl req -new -x509 -days 3654 -nodes -config "$openssl_conf" -out "$CONFDPEM" -keyout "$CONFDPEM" || RC="1"
 openssl gendh -rand $tmp_opsiconfd_rand 512 >> "$CONFDPEM" || RC="1"
 openssl x509 -subject -dates -fingerprint -noout -in "$CONFDPEM" || RC="1"
 rm -f "$rand_conf" "$openssl_conf"
 return "$RC"
}

create_certs(){
 local RC="0"
 # lmn7
 if [ -s "$OPSIKEY" -a -s "$OPSICERT" ]; then
   cat "$OPSIKEY" "$OPSICERT" > "$CONFDPEM" || RC="1"
 else
   # lmn62
   rm -f /etc/ssh/ssh_host*key*
   ssh-keygen -t dsa -f /etc/ssh/ssh_host_dsa_key -N '' || RC="1"
   ssh-keygen -t rsa -f /etc/ssh/ssh_host_rsa_key -N '' || RC="1"
   create_opsi_cert || RC="1"
 fi
 # opsi
 return "$RC"
}

# functions end


# sets host and domainname
# hostname
if [ "$MYFQDN" != "opsi.$domainname" ]; then
 cert_ok=no
 set_fqdn || RC="1"
fi

# set ip in config.ini
set_ip || RC="1"

# repair opsi permissions
opsi-setup --set-rights "$OPSISYSDIR" || RC="1"
opsi-setup --set-rights "$OPSICONFIGDIR" || RC="1"

# create ssh and host certs
if [ -n "$first" -o "$cert_ok" != "yes" ]; then
 create_certs || RC="1"
fi

# opsi setup finally
opsi-setup --auto-configure-samba || RC="1"
opsi-setup --init-current-config || RC="1"
service opsipxeconfd stop
service opsiconfd stop
rm -f /var/run/opsi*/*.pid
service opsiconfd start || RC="1"
service opsipxeconfd start || RC="1"


share/network.sh
#
# linuxmuster-opsi-network
#
# thomas@linuxmuster.net
# 20170718
#

# handle network interface to use
if [ -z "$iface" ]; then
  # get list of ethernet interfaces
  ifaces=( $(ip link show | grep -B1 'link/ether' | grep ^[0-9] | awk '{ print $2 }' | awk -F\: '{ print $1 }') )
  if [ ${#ifaces[@]} -eq 0 ]; then
    echo "No ethernet interface found!"
    usage
  elif [ ${#ifaces[@]} -eq 1 ]; then
    # only one interface found, use it
    iface="$ifaces"
  else
    # more than one interfaces
    i="$(echo ${ifaces[@]})"
    while true; do
      read -p "Enter network interface to use [$i]: " -a iface
      echo $i | grep -qw $iface && break
    done
  fi
fi
# test if interface is already configured
res="$(grep -r ^iface /etc/network/* | grep -v ^"$IFACES_TGT" | grep -qw "$iface" | awk -F\: '{ print $1 }' | head -1)"
if [ -n "$res" ]; then
  echo "Interface $iface is already configured in $res."
  exit 1
fi

# save opsiip so they do not get overwritten by settings
opsiip_new="$opsiip"

# read previous setup values if present
if [ -s "$SETTINGS" ]; then
  source "$SETTINGS"
  # restore given opsiip
  [ -n "$opsiip_new" ] && opsiip="$opsiip_new"
fi

# handle ip address
if [ -z "$opsiip_new" ]; then
  # set default value
  [ -z "$opsiip" ] && opsiip="10.16.1.2"
  while true; do
    read -e -i "$opsiip" -p "Enter ip address: " input
    opsiip="${input:-$opsiip}"
    validip $opsiip && break
  done
fi

# handle netmask
if [ -z "$netmask" ]; then
  [ -n "$internalnet" -a -n "$broadcast" ] && netmask="$(ipcalc -b $(ipcalc $internalnet - $broadcast | tail -1) | grep ^Netmask | awk '{ print $2 }')"
  # set default value
  [ -z "$netmask" ] && netmask="255.240.0.0"
  while true; do
    read -e -i "$netmask" -p "Enter netmask: " input
    netmask="${input:-$netmask}"
    validmask $netmask && break
  done
fi

# handle gateway address
if [ -z "$gateway" ]; then
  gateway="$ipcopip"
  # set default value
  [ -z "$gateway" ] && gateway="10.16.1.254"
  while true; do
    read -e -i "$gateway" -p "Enter gateway address: " input
    gateway="${input:-$gateway}"
    validip $gateway && break
  done
fi

# handle nameserver address
if [ -z "$nameserver" ]; then
  nameserver="$serverip"
  # set default value
  [ -z "$nameserver" ] && nameserver="10.16.1.1"
  while true; do
    read -e -i "$nameserver" -p "Enter nameserver address: " input
    nameserver="${input:-$nameserver}"
    validip $nameserver && break
  done
fi

# write interface file
ifdown "$iface"
sed -e "s|@@iface@@|$iface|g
        s|@@opsiip@@|$opsiip|
        s|@@netmask@@|$netmask|
        s|@@gateway@@|$gateway|
        s|@@nameserver@@|$nameserver|" "$IFACES_TPL" > "$IFACES_TGT"
# change dns-search entry if domainname is set, default is localhost.localdomain
[ -n "$domainname" ] && sed -i "s|dns-search .*|dns-search $domainname|" "$IFACES_TGT"
ifup "$iface"


share/prepare.sh
#
# linuxmuster-opsi-prepare
#
# thomas@linuxmuster.net
# 20190311
#

# upgrade and install necessary pkgs if not done before
pkgs_installed || dist_upgrade | tee -a "$LOGFILE"

# copy initial configs

# sudoers
if ! grep -q ^opsiconfd "$SUDOERS_TGT"; then
 echo "Updating $SUDOERS_TGT."
 cp "$SUDOERS_TPL" "$SUDOERS_TGT" || RC="1"
 chmod 440 "$SUDOERS_TGT"
fi

# opsi configs
cp "$SHAREDIR/dbp.repo" "$OPSIREPOHOOKS"
cp "$SHAREDIR/dispatch.conf" "$OPSIBCKNDMGR"
sed -i 's|^autoInstall .*|autoInstall = true|' "$OPSIWINREPO"

# opsi setup stuff
opsi-setup --auto-configure-samba || RC="1"
opsi-setup --init-current-config || RC="1"
opsi-setup --set-rights || RC="1"
service opsipxeconfd stop
service opsiconfd stop
rm -f /var/run/opsi*/*.pid
service opsiconfd start || RC="1"
service opsipxeconfd start || RC="1"
[ "$RC" = "0" ] || bailout "Opsi setup error!"

# opsiadmin user
if ! id "$ADMINUSER" &> /dev/null; then
 echo
 echo "Creating $ADMINUSER account ..."
 useradd -c "OPSI admin user" -g "$ADMINGROUP" -G adm,cdrom,sudo,dip,plugdev,sambashare,pcpatch -m -s /bin/bash "$ADMINUSER" || bailout "Cannot create $ADMINUSER!"
fi

echo
echo "The OPSI system has been successfully prepared."
echo "Please invoke now on the server this command:"
echo "# linuxmuster-opsi --setup --first"


share/wsimport.sh
#
# sync workstations with opsi management
#
# thomas@linuxmuster.net
# 20200418
#


# read linuxmuster.net environment
. "$SETTINGS" || exit 1


# create temp dir for run scripts
RUNDIR="/var/tmp/wsimport.$$"
[ -e "$RUNDIR" ] && rm -rf "$RUNDIR"
mkdir -p "$RUNDIR"
if [ ! -d "$RUNDIR" ]; then
 bailout "Cannot create temporary directory $RUNDIR!"
fi


# functions begin

# generates runfiles
generate_runfile(){
 local action="$1"
 local pcname="$2"
 local mac="$3"
 local ip="$4"
 local RC="0"
 local cmd
 local runfile
 case "$action" in
  create)
   cmd="opsi-admin -d method host_createOpsiClient $pcname.$domainname null $pcname \"\" $mac $ip"
   runfile="$RUNDIR/01_${action}_$pcname" ;;
  remove)
   cmd="opsi-admin -d method host_delete $pcname.$domainname"
   runfile="$RUNDIR/00_${action}_$pcname" ;;
  *) ;;
 esac
 [ -z "$cmd" -o -z "$runfile" ] && return
 echo -n "Generating $action runfile for $pcname ..."
 cat > "$runfile" <<EOF
#!/bin/sh
RC="0"
echo -n "Running $action opsi client $pcname.$domainname ..."
$cmd || RC="1"
if [ "\$RC" = "0" ]; then echo " OK."; else echo " Failed!"; fi
exit "\$RC"
EOF
 RC="$?"
 chmod +x "$runfile" || RC="1"
 if [ "$RC" = "0" ]; then echo " OK."; else echo " Failed!"; fi
 return "$RC"
}

# functions end


# iterate over client names
grep ^[a-z0-9] "$WIMPORTDATA" | while read data; do

 # filter out clients not scheduled for opsi
 pxe="$(echo "$data" | awk -F\; '{ print $11 }')"
 [ "$pxe" = "2" -o "$pxe" = "3" ] || continue

 pcname="$(echo "$data" | awk -F\; '{ print $2 }')"

 # not opsi himself
 [ "$pcname" = "opsi" ] && continue

 mac="$(echo "$data" | awk -F\; '{ print $4 }' | tr A-Z a-z)"
 ip="$(echo "$data" | awk -F\; '{ print $5 }' | tr A-Z a-z)"
 # dynamic ip client support
 [ "$ip" = "dhcp" ] && ip=""
 inifile="$OPSICLIENTSDIR/$pcname.$domainname.ini"
 # remove opsi clients whose mac or ip address has changed
 # note: this client has to be synchronized with linbo afterwards to get new key and current opsi status
 if [ -e "$inifile" ]; then
  RCTMP="0"
  CHANGED=""
  if ! grep ^hardwareaddress "$inifile" | grep -qi "$mac"; then
   echo "MAC address of client $pcname has changed!"
   CHANGED="yes"
 # only if ip address is set (dynamic ip  client support)
 elif [ -n "$ip" ] && ! grep ^ipaddress "$inifile" | grep -q "$ip"; then
   echo "IP address of client $pcname has changed!"
   CHANGED="yes"
  fi
  if [ -n "$CHANGED" ]; then
   generate_runfile remove "$pcname" || RCTMP="1"
   generate_runfile create "$pcname" "$mac" "$ip" || RCTMP="1"
   [ "$RCTMP" = "0" ] || RC="1"
  fi
 fi

 # test for client not yet imported in opsi and generate a runfile if he has to be imported
 RCTMP="0"
 grep -qwi ^"$pcname" "$PCKEYS" || generate_runfile create "$pcname" "$mac" "$ip" ; RCTMP="$?"
 [ "$RCTMP" = "0" ] || RC="1"

done

# remove clients which are not in workstations file and generate runfiles for them
pcs=$(grep ^[a-zA-Z0-9] "$WIMPORTDATA" | awk -F\; '{ print $2, $11 }' | grep " [23]" | awk '{ print $1 }')
for pcname in $(grep ^[a-zA-Z0-9] "$PCKEYS" | awk -F \. '{ print $1 }' | tr A-Z a-z | grep -vw ^opsi); do

 RCTMP="0"
 echo "$pcs" | grep -qwi "$pcname" || generate_runfile remove "$pcname" ; RCTMP="$?"
 [ "$RCTMP" = "0" ] || RC="1"

done

# finally run generated scripts
if ls "$RUNDIR"/* &> /dev/null; then
 run-parts "$RUNDIR" || RC="1"
fi
rm -rf "$RUNDIR"


share/opsiSetup.py
#!/usr/bin/python3
#
# opsiSetup
# thomas@linuxmuster.net
# 20190314
#

import configparser
import os
import sys

# arguments
clientname = sys.argv[1]
try:
    forcelist = sys.argv[2]
except:
    forcelist = None

# path to client ini
inidir = '/var/lib/opsi/config/clients'
inifile = inidir + '/' + clientname + '.ini'
bakfile = inifile + '.bak'

if not os.path.isfile(inifile):
    print('Inifile not found!')
    sys.exit(1)
if not os.path.isfile(bakfile):
    print('Old inifile not found!')
    sys.exit(1)

# read client inis
newini = configparser.ConfigParser(inline_comment_prefixes=('#', ';'))
backupini = configparser.ConfigParser(inline_comment_prefixes=('#', ';'))
newini.read(inifile)
backupini.read(bakfile)

# get old products and versions and collect them in an array
setup_products = []
for (product, product_state) in backupini.items('localboot_product_states'):
    # skip not installed products
    if 'installed' not in product_state:
        continue
    bak_version = backupini.get(product + '-state', 'productversion')
    try:
        new_version = newini.get(product + '-state', 'productversion')
    except:
        setup_products.append(product)
        continue
    if bak_version != new_version:
        setup_products.append(product)
        continue
    # add forced products
    if forcelist not None:
        for item in forcelist.split(','):
            if product == item:
                setup_products.append(product)
                break

# change action to setup for collected products
if len(setup_products) > 0:
    print('Changing action to setup for:')
    for product in setup_products:
        print('# ' + product)
        backupini.set('localboot_product_states', product, 'installed:setup')
else:
    print('Nothing to do.')
    sys.exit(0)

# write changed backup file
print('Writing changed client ini.')
with open(bakfile, 'w') as outfile:
    backupini.write(outfile)

# move changed file in place
os.system('mv ' + bakfile + ' ' + inifile)
os.system('opsi-set-rights ' + inidir)


share/opsi41upgrade.sh
#!/bin/sh
#
# Upgrade opsi to 4.1 and ubuntu to 18.04
# thomas@linuxmuster.net
# 20190311
#

if [ -d /etc/opsi/package-updater.repos.d ]; then
  echo "Opsi is already upgraded to version 4.1!"
  exit 1
fi

if [ "$1" = "-h" -o "$1" = "--help" ]; then
  echo "Usage: $0 \"<opsiadmin password>\""
  exit 0
fi

# get opsiadmin password from command line
password="$1"

distupgrade(){
  RC="0"
  export DEBIAN_FRONTEND=noninteractive
  apt update || RC="1"
  echo -e '\n\n\n' | apt-get -y dist-upgrade || RC="1"
  apt clean || RC="1"
  apt-get -y autoremove || RC="1"
  return "$RC"
}

opsisetup(){
  RC="0"
  opsi-setup --auto-configure-samba || RC="1"
  opsi-setup --init-current-config || RC="1"
  opsi-setup --set-rights || RC="1"
  service opsipxeconfd stop
  service opsiconfd stop
  rm -f /var/run/opsi*/*.pid
  service opsiconfd start || RC="1"
  service opsipxeconfd start || RC="1"
  return "$RC"
}

RC="0"
rm -f /etc/apt/sources.list.d/lmn7.list
sed -i 's|xenial|bionic|g' /etc/apt/sources.list
sed -e 's|@@issue@@|18.04|g' /usr/share/linuxmuster-opsi/opsi.list > /etc/apt/sources.list.d/opsi.list
distupgrade || RC="1"
cp /usr/share/linuxmuster-opsi/dbp.repo /etc/opsi/package-updater.repos.d
cp /usr/share/linuxmuster-opsi/dispatch.conf /etc/opsi/backendManager
opsisetup || RC="1"
if [ -n "password" ]; then
  linuxmuster-opsi --password="$password"
else
  linuxmuster-opsi --password || RC="1"
fi

if [ "$RC" = "0" ]; then
  echo "Upgrade done, please reboot!"
else
  echo "Finished with errors!"
fi
